text/0066-archival-nodes.md

```diff
@@ -0,0 +1,86 @@
+
+This CEP outlines the design of archival nodes, that is, nodes keeping the complete history of the Casper blockchain.
+
+## Motivation
+
+[motivation]: #motivation
+
+The Casper blockchain, like all blockchains, is rapidly growing, and requires more and more storage space to be kept in its entirety. This is very wasteful, as only a small number of most recent blocks are strictly required for regular operations. It is thus tempting to simply discard the blocks that are no longer relevant, but this would mean losing almost all of the network's history, which could turn out to be catastrophic if something goes wrong, and would go against the idea of a blockchain being a public ledger of all operations ever executed on it.
+
+There is a middle ground here, though - instead of having all nodes keep the whole history, or having all nodes discard older blocks, we could have most of the nodes only keeping what is relevant for day-to-day operations, and only a small number of nodes keeping the complete history of the network. These history keepers would be called the "archival nodes".
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Until Fast Sync is introduced, all nodes are effectively archival nodes, as every time a node joins the network, it has to sync the whole blockchain since genesis. Thus, once it becomes able to participate in the network, it already has all the history stored locally.
+
+Fast Sync is going to change that picture. Once it starts being the method used for syncing, the joining nodes will only download some of the switch block headers and the global state from a recent block, ignoring a huge part of the network's history. This means that if a fast syncing node didn't have all the history already, it won't have at least some of it afterwards, either.
+
+Since Fast Sync has to be able to download full data of at least a single block, we propose for the archival nodes to use Fast Sync capabilities to sync block by block, from genesis up to the most recent blocks.
+
+The process will be analogous to the pre-fast-sync Linear Chain Sync:
+
+1. Walk back from trusted hash back to genesis.
+2. Download blocks one at a time starting from genesis upwards.
+    - The difference here would be that instead of executing the blocks, we would just download the global state trie from our peers.
+3. When we're in the current protocol version and can't fetch any more blocks, stop.
+
+After completing this process, the node will have the complete chain and all historical global state tries.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Going into more details of the process explained above, it would look like this:
+
+1. Download the header corresponding to the trusted hash.
+2. While the current header's height isn't 0, download the parent header by hash.
+3. When height 0 is reached, read the genesis set of validators for validating finality signatures.
```

Comment:
We don't actually have to validate the finality signatures before the trusted block. (And signatures of unbonded validators shouldn't be trusted anymore anyway.)

Reply:
Per our call, this will get rewritten as a top-down sync.